src/GenesisApp.cpp

```diff
@@ -41,13 +43,124 @@ GenesisApp::GenesisApp()
 ////////////////////////////////////////////////////////////////////////
 	: BApplication(GENESIS_APP_SIG)
 {
+	fMainWindowsList = new BObjectList<GenesisWindow>(5);
+	fQuitConfirmed = false;
+	fPrefWindow = NULL;
+}
+
+////////////////////////////////////////////////////////////////////////
+void GenesisApp::NewWindow()
+////////////////////////////////////////////////////////////////////////
+{
+	GenesisWindow* w;
 
+	w = new GenesisWindow();
+	w->Show();
+
+	fMainWindowsList->AddItem(w);
+}
+
+////////////////////////////////////////////////////////////////////////
+void GenesisApp::OpenPreferences(GenesisWindow *w = NULL)
+////////////////////////////////////////////////////////////////////////
+{
+	if (fPrefWindow == NULL)
+	{
+		fPrefWindow = new GenesisPreferencesWindow();
+		fPrefWindow->Show();
+	}
+	fPrefWindow->Activate();
+
+	if (w != NULL)
+	{
+		fPrefWindow->CenterIn(w->Frame());
+	}
+}
+
+////////////////////////////////////////////////////////////////////////
+int GenesisApp::GetMainWinCount()
+////////////////////////////////////////////////////////////////////////
+{
+	return fMainWindowsList->CountItems();
+}
+
+////////////////////////////////////////////////////////////////////////
+bool GenesisApp::QuitConfirmed()
+////////////////////////////////////////////////////////////////////////
+{
+	return fQuitConfirmed;
 }
 
 ////////////////////////////////////////////////////////////////////////
 void GenesisApp::ReadyToRun()
 ////////////////////////////////////////////////////////////////////////
 {
-	fWindow = new GenesisWindow();
-	fWindow->Show();
+	NewWindow();
+}
+
+////////////////////////////////////////////////////////////////////////
+bool GenesisApp::QuitRequested()
+////////////////////////////////////////////////////////////////////////
+{
+	if (fMainWindowsList->CountItems() > 0 && SETTINGS->GetAskOnExit())
+	{
+		BAlert *myAlert = new BAlert("Quit Genesis?","Do you really want to quit?","No","Quit",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_WARNING_ALERT);
```

Comment:
Always a good idea to "decouple" the labels of the buttons from the alert text:
```suggestion
		BAlert *myAlert = new BAlert("Quit Genesis?","Do you really want to quit?","Cancel","Quit",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_WARNING_ALERT);
```